File: smallpict-go/crypto_test.go
package smallpict

import (
	"testing"
)

func TestCryptoUtilities(t *testing.T) {
	if EmptySHA256 != "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" {
		t.Fatalf("unexpected EmptySHA256 constant: %s", EmptySHA256)
	}

	if h := SHA256Hex(nil); h != EmptySHA256 {
		t.Fatalf("expected empty SHA256, got: %s", h)
	}

	if h := SHA256Hex([]byte("hello world")); h != "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9" {
		t.Fatalf("unexpected SHA256 for 'hello world': %s", h)
	}

	stringToSign := BuildStringToSign("POST", "/v1/optimize", "1716301234", EmptySHA256)
	expectedSTS := "POST\n/v1/optimize\n1716301234\ne3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
	if stringToSign != expectedSTS {
		t.Fatalf("unexpected string to sign: %s", stringToSign)
	}

	sig := HMACSHA256Hex("sec_test_secret_key_123", stringToSign)
	if len(sig) != 64 {
		t.Fatalf("expected 64 character signature, got len %d: %s", len(sig), sig)
	}
}

func TestSanitizeMessage(t *testing.T) {
	raw := "Failed key sp_live_1234567890abcdef1234567890abcdef with sec_secret123456"
	sanitized := SanitizeMessage(raw)

	if sanitized == raw {
		t.Fatalf("expected sanitization, got raw: %s", sanitized)
	}
	if len(sanitized) == 0 {
		t.Fatal("expected non-empty sanitized string")
	}
}

File: smallpict-rust/tests/crypto_test.rs
use smallpict::crypto::{self, EMPTY_SHA256};
use smallpict::errors::{sanitize_message, SecretRedacted};

#[test]
fn test_empty_sha256_constant() {
    assert_eq!(
        EMPTY_SHA256,
        "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    );
    assert_eq!(crypto::sha256_hex(&[]), EMPTY_SHA256);
}

#[test]
fn test_sha256_hex() {
    assert_eq!(
        crypto::sha256_hex(b"hello world"),
        "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9"
    );
}

#[test]
fn test_hmac_sha256_hex() {
    let string_to_sign =
        crypto::build_string_to_sign("POST", "/v1/optimize", "1716301234", EMPTY_SHA256);
    let expected_sts =
        "POST\n/v1/optimize\n1716301234\ne3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
    assert_eq!(string_to_sign, expected_sts);

    let sig = crypto::hmac_sha256_hex("sec_test_secret_key_123", &string_to_sign);
    assert_eq!(sig.len(), 64);
}

#[test]
fn test_sanitize_message() {
    let raw = "Failed key sp_live_1234567890abcdef1234567890abcdef with sec_secret123456";
    let sanitized = sanitize_message(raw);

    assert!(!sanitized.contains("sp_live_1234567890abcdef1234567890abcdef"));
    assert!(!sanitized.contains("sec_secret123456"));
    assert!(sanitized.contains("sp_live_12...cdef"));
    assert!(sanitized.contains("***REDACTED***"));
}

#[test]
fn test_secret_redacted_wrapper() {
    let secret = SecretRedacted("sp_live_1234567890abcdef1234567890abcdef".to_string());
    assert_eq!(format!("{:?}", secret), "[REDACTED]");
    let disp = format!("{}", secret);
    assert!(!disp.contains("sp_live_1234567890abcdef1234567890abcdef"));
    assert!(disp.contains("sp_live_12...cdef"));
}
